seq2struct/utils/parallelizer.py

```diff
@@ -0,0 +1,114 @@
+
+
+from abc import ABC, abstractmethod
+
+import multiprocessing
+import tqdm
+
+class Parallelizer(ABC):
+    def __init__(self, nproc, batch_size):
+        """
+        Create a Parallelizer
+
+        Args
+            nproc: the total number of processes to use
+            batch_size: the size of a single chunk of elements to be sent to a process
+        """
+        assert isinstance(nproc, int) and nproc >= 1
+        assert isinstance(batch_size, int) and batch_size >= 1
+        self.nproc = nproc
+        self.batch_size = batch_size
+
+    @abstractmethod
+    def _map(self, fn, args):
+        """
+        Map the given function over the arguments, and yields each of the results.
+
+        This function should behave functionally identically to the builtin `map`.
+
+        This function should assume no batching, sending each of the arguments individually
+            to a process. It also does not need to handle any progress-bar logic.
+        """
+
+    @abstractmethod
+    def _parallel_process_bar_class(self):
+        """
+        Return a class with an interface similar to tqdm.tqdm, whose update function
+            can be shipped to individual processes.
+        """
+
+    def map_batched(self, fn, args):
+        """
+        Run the function fn on each of the args, in parallel, and yields each of
+            the results.
+
+        This function should behave functionally identically to the builtin `map`
+
+        Args
+            fn: the function to run
+            args: an iterable of arguments to run it on
```

Comment:
Better to clarify that `args` is an iterable whose elements each make up one argument to a function, since `args` in Python usually means one set of arguments for a single function call.